src/coord/src/sink_connector.rs

```diff
@@ -118,34 +118,66 @@ async fn get_latest_ts(
     // Seek to end-1 offset
     let mut tps = TopicPartitionList::new();
     tps.add_partition(consistency_topic, partition);
-    tps.set_partition_offset(consistency_topic, partition, Offset::OffsetTail(1))?;
+    tps.set_partition_offset(consistency_topic, partition, Offset::Beginning)?;
 
-    consumer
-        .assign(&tps)
-        .with_context(|| format!("Error seeking in consistency topic {}", consistency_topic))?;
+    consumer.assign(&tps).with_context(|| {
+        format!(
+            "Error seeking in consistency topic {}:{}",
+            consistency_topic, partition
+        )
+    })?;
+
+    // We scan from the beginning and see if we can find an END record. We have
+    // to do it like this because Kafka Control Batches mess with offsets. We
+    // thereforecannot simply take the last offset from the back and expect an
+    // END message there. With a transactional producer, the OffsetTail(1) will
+    // not point to an END message but a control message. With aborted
+    // transactions, there might even be a lot of garbage at the end of the
+    // topic or in between.
+
+    let mut latest_message = None;
+    while let Some(message) = get_next_message(consumer, timeout)? {
+        latest_message = Some(message);
+    }
 
-    // Read the end-1 offset message
-    let m = match get_next_message(consumer, timeout)? {
+    let latest_message = match latest_message {
```

Comment:
I find this sort of pattern where you either want to:

do a bunch of logic if something is None,
OR just unwrap if something is Some(_) is a lot nicer / less nested if you make that explicit e.g.

```
diff --git a/src/coord/src/sink_connector.rs b/src/coord/src/sink_connector.rs
index 531fcea1e..b32db7632 100644
--- a/src/coord/src/sink_connector.rs
+++ b/src/coord/src/sink_connector.rs
@@ -140,30 +140,29 @@ async fn get_latest_ts(
         latest_message = Some(message);
     }

-    let latest_message = match latest_message {
-        None => {
-            // fetch watermarks to distinguish between a timeout reading end-1 and an empty topic
-            match consumer.fetch_watermarks(consistency_topic, 0, timeout) {
-                Ok((lo, hi)) => {
-                    if hi == 0 {
-                        return Ok(None);
-                    } else {
-                        bail!(
-                            "uninitialized consistency topic {}:{}, lo/hi: {}/{}",
-                            consistency_topic,
-                            partition,
-                            lo,
-                            hi
-                        );
-                    }
-                }
-                Err(e) => {
-                    bail!("Failed to fetch metadata while reading from consistency topic, will retry. Error was {}", e);
+    if latest_message.is_none() {
+        // fetch watermarks to distinguish between a timeout reading end-1 and an empty topic
+        match consumer.fetch_watermarks(consistency_topic, 0, timeout) {
+            Ok((lo, hi)) => {
+                if hi == 0 {
+                    return Ok(None);
+                } else {
+                    bail!(
+                        "uninitialized consistency topic {}:{}, lo/hi: {}/{}",
+                        consistency_topic,
+                        partition,
+                        lo,
+                        hi
+                    );
                 }
             }
+            Err(e) => {
+                bail!("Failed to fetch metadata while reading from consistency topic, will retry. Error was {}", e);
+            }
         }
-        Some(message) => message,
-    };
+    }
+
+    let latest_message = latest_message.expect("known to exist");

     // the latest valid message should be an END message. If not things have
     // gone wrong!
```

Reply:
thanks! I'll use that.